drivers/regulator/pmic_shell.c

```diff
@@ -0,0 +1,169 @@
+{
+	int vol;
+
+	if (pmic_dev == NULL) {
+		shell_print(sh, "No PMIC device has been selected");
+		return -ENODEV;
+	}
+	if (!device_is_ready(pmic_dev)) {
+		shell_print(sh, "PMIC device not ready");
+		return -ENODEV;
+	}
+	if (argc != 2) {
+		shell_print(sh, "usage: %s [limit (mV)]", argv[0]);
+		shell_print(sh, "A voltage must be provided");
+		return -EINVAL;
+	}
+	vol = atoi(argv[1]) * 1000;
+	shell_print(sh, "Setting to %d uV", vol);
+	if (regulator_set_voltage(pmic_dev, vol, MAX_REGULATOR_VOLTAGE) < 0) {
+		shell_print(sh, "Failed to set voltage");
+		return -EACCES;
+	}
+	vol = regulator_get_voltage(pmic_dev);
+	if (vol < 0) {
+		shell_print(sh, "Failed to read voltage");
+		return -EACCES;
+	}
+	shell_print(sh, "Set voltage to %d uV", vol);
+	return 0;
+}
+
+static int cmd_pmic_set_ilim(const struct shell *sh, size_t argc, char **argv)
+{
+	int cur;
+
+	if (pmic_dev == NULL) {
+		shell_print(sh, "No PMIC device has been selected");
+		return -ENODEV;
+	}
+	if (!device_is_ready(pmic_dev)) {
+		shell_print(sh, "PMIC device not ready");
+		return -ENODEV;
+	}
+	if (argc != 2) {
+		shell_print(sh, "usage: %s [limit (mA)]", argv[0]);
+		shell_print(sh, "A current limit must be provided");
+		return -EINVAL;
+	}
+	cur = atoi(argv[1]) * 1000;
+	shell_print(sh, "Setting to %d uA", cur);
+	if (regulator_set_current_limit(pmic_dev, cur, 7000000) < 0) {
+		shell_print(sh, "Failed to set current limit");
+		return -EACCES;
+	}
+	cur = regulator_get_current_limit(pmic_dev);
+	if (cur < 0) {
+		shell_print(sh, "Failed to read current limit");
+		return -EACCES;
+	}
+	shell_print(sh, "Set current limit to %d uA", cur);
+	return 0;
+}
+
+static int cmd_pmic_set_dev(const struct shell *sh, size_t argc, char **argv)
+{
+	const struct device *tmp;
+
+	if (argc != 2) {
+		shell_print(sh, "usage: %s [device name]", argv[0]);
+		shell_print(sh, "A device must be provided");
+		return -EINVAL;
+	}
+	tmp = device_get_binding(argv[1]);
+	if (tmp == NULL) {
+		shell_print(sh, "Device %s not found", argv[2]);
+		return -ENODEV;
+	}
+	pmic_dev = tmp;
+	shell_print(sh, "Device set to %s", pmic_dev->name);
+	return 0;
+}
+
+static int cmd_list_devices(const struct shell *sh, size_t argc, char **argv)
+{
+	for (int i = 0; i < ARRAY_SIZE(supported_pmics); i++) {
+		shell_print(sh, "%s", supported_pmics[i]->name);
+	}
+	return 0;
+}
+
+
+SHELL_STATIC_SUBCMD_SET_CREATE(pmic_set,
+	SHELL_CMD(enable, NULL, "Enable buck regulator", cmd_pmic_en),
+	SHELL_CMD(disable, NULL, "Disable buck regulator", cmd_pmic_dis),
+	SHELL_CMD(set_vol, NULL, "Set voltage (in mV)", cmd_pmic_set_vol),
+	SHELL_CMD(set_current, NULL, "Set current limit( in mA)",
+		cmd_pmic_set_ilim),
+	SHELL_CMD(set_device, NULL, "Set regulator device", cmd_pmic_set_dev),
+	SHELL_CMD(list_devices, NULL, "List enabled regulators",
+		cmd_list_devices),
+	SHELL_SUBCMD_SET_END
+);
+SHELL_CMD_REGISTER(pmic, &pmic_set, "PMIC Management", NULL);
```

Comment:
I would prefer a generic top-level `regulator` command with `enable` and `disable` commands - and a conditional `pmic` subcommand for the PMIC specific commands. This will make the shell module useful for non-PMIC regulators as well.

Reply:
I've updated the shell to use this approach, so that generic regulators will work